### PR TITLE
[WIP] add RoutedDecoderIterDataset prototype

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48849 webdataset ExpandUrlMacroIterDataset protype
* **#48295 [WIP] add RoutedDecoderIterDataset prototype**
* #47831 [WIP] web dataset prototype

This is a prototype for RoutedDecoderIterDataset, partially based on Nvidia's webdataset

Outstanding queston:
We pass stream handle cross dataset, but when decoding, we have to read all the content out of stream. This means the stream handle will no longer be useable after decode. (Assuming the stream handle is not seekable and this is the case for web stream). Since we use yield which means no matter how many layers, we just process one item each time, given this, should we pass bytes data instead of stream handle? This will be much easier to maintain and should not take much memory. 
Need to discuss.

Differential Revision: [D25115107](https://our.internmc.facebook.com/intern/diff/D25115107)

### CODE REVIEW

File: torch/utils/data/decoder.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        except ValueError:
+            return None
+
+    if extension in "json jsn":
+        return json.loads(data)
+
+    if extension in "pyd pickle".split():
+        return pickle.loads(data)
+
+    if extension in "pth".split():
```
Comment (vadimkantorov) [inline torch/utils/data/decoder.py:40]: `.pth` is not the best extension to standardize as PyTorch model extension: https://github.com/pytorch/pytorch/issues/14864

`.pt` is safer

File: torch/utils/data/decoder.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+
+################################################################
+# handle basic datatypes
+################################################################
+
+
+def basichandlers(key, data):
+
+    extension = re.sub(r".*[.]", "", key)
```
Comment (SsnL) [inline torch/utils/data/decoder.py:23]: os.path.splitext ?

File: torch/utils/data/_dataset.py [hunk 3]
```diff
@@ -133,3 +137,34 @@ def __iter__(self):
...
         return self.length
+
+
+# Dataset for decoding data in different format (eg, json, png, etc)
+# Yield tuple(pathname, decoded object)
+# args:
+#    input_dataset - iterable with tuple(pathname, binary stream or bytes)
+#    decoders      - user passed in decoders/handlers, if None, a set of default decoders will be initialized
+#    length        - a nominal length of the dataset
+class RoutedDecoderIterDataset(IterDataset):
```
Comment (SsnL) [inline torch/utils/data/_dataset.py:148]: We sure that we want to expose `IterableDataset` and `XIterDataset`? Sounds extremely error-prune to me.

Comment (glaringlee) [reply torch/utils/data/_dataset.py:148]: @SsnL This is a prototype which is used for further design discussion on Dataset. Currently, the high level design is to expose different kinds of Dataset, yes, XIterDataset, and maybe further XMapDataset. Everything is dataset, sampler, shuffler, sharding, etc.  This is could be more error-prune, need careful design and implementation, but we do have lots of benefit on decoupling the current dataset into smaller pieces. We are still working on a design doc, you guys can comment once it is ready. cc @VitalyFedyunin

Comment (SsnL) [reply torch/utils/data/_dataset.py:148]: I was just mostly talking about the inconsistent naming.

Comment (glaringlee) [reply torch/utils/data/_dataset.py:148]: Naming is under discussion as well.

**Comment by glaringlee** [comment]
move to #48849 due to stack messed up.

**Pull request closed without merging**